File: production/client/src/main/scala/kpn/client/filter/BooleanFilter.scala
// Migrated to Angular: boolean-filter.ts
package kpn.client.filter

import japgolly.scalajs.react.Callback

class BooleanFilter[T](
  name: String,
  val criterium: Option[Boolean],
  val booleanPropertyAccessor: (T) => Boolean,
  all: Callback = Callback.empty,
  yes: Callback = Callback.empty,
  no: Callback = Callback.empty
) extends Filter[T](name) {

  def passes(element: T): Boolean = {
    if(criterium.isEmpty) {
      true
    }
    else {
      if(criterium.get) {
        booleanPropertyAccessor(element)
      }
      else {
        ! booleanPropertyAccessor(element)
      }
    }
  }

  def filterOptions(allFilters: Filters[T], elements: Seq[T]): Option[FilterOptionGroup] = {

    val filteredElements = allFilters.filterExcept(elements, this)
    val (yesElements, noElements) = filteredElements.partition(booleanPropertyAccessor)
    val active = filteredElements.nonEmpty && yesElements.nonEmpty && noElements.nonEmpty

    if (active) {
      val allOption = FilterOption(
        "all",
        filteredElements.size,
        criterium.isEmpty,
        all
      )

      val yesOption = FilterOption(
        "yes",
        yesElements.size,
        criterium.contains(true),
        yes
      )

      val noOption = FilterOption(
        "no",
        noElements.size,
        criterium.contains(false),
        no
      )

      Some(FilterOptionGroup(name, allOption, yesOption, noOption))
    }
    else {
      None
    }
  }
}

File: frontend/src/app/shared/kpn/filter/boolean-filter.ts
import { Filter } from './filter';
import { FilterOption } from './filter-option';
import { FilterOptionGroup } from './filter-option-group';
import { Filters } from './filters';

export class BooleanFilter<T> extends Filter<T> {
  constructor(
    name: string,
    private readonly criterium: boolean,
    private readonly getter: (row: T) => boolean,
    private readonly all: () => void,
    private readonly yes: () => void,
    private readonly no: () => void
  ) {
    super(name);
  }

  passes(element: T): boolean {
    if (this.criterium == null) {
      return true;
    }
    if (this.criterium) {
      return this.getter(element);
    }
    return !this.getter(element);
  }

  filterOptions(allFilters: Filters<T>, elements: T[]): FilterOptionGroup {
    const filteredElements = allFilters.filterExcept(elements, this);
    const yesElements = filteredElements.filter((e) => this.getter(e));
    const noElements = filteredElements.filter((e) => !this.getter(e));
    const active = filteredElements.length > 0 && yesElements.length > 0 && noElements.length > 0;

    if (active) {
      const allOption = new FilterOption(
        'all',
        filteredElements.length,
        this.criterium == null,
        this.all
      );

      const yesOption = new FilterOption(
        'yes',
        yesElements.length,
        this.criterium === true,
        this.yes
      );

      const noOption = new FilterOption('no', noElements.length, this.criterium === false, this.no);

      return new FilterOptionGroup(this.name, allOption, yesOption, noOption);
    }
    return null;
  }
}
